packages/common/container-definitions/src/audience.ts

```diff
@@ -67,4 +73,25 @@ export interface IAudience extends EventEmitter {
 	 * does not technically have a clientId tied to an active connection to the service.
 	 */
 	getMember(clientId: string): IClient | undefined;
+
+	/**
+	 * Returns this client's clientId, if it exists. undefined if this client never connected to ordering service.
+	 * Whenever this property changes, "clientIdChanged" event is fired on this object.
+	 * Wheeever clients loses connection and reconnects, it will raise "connected" event at various API surfaces.
+	 *
+	 * It's guranteed that such events and change of current clientId happens at the same time (syncronously, one after another).
+	 * That said, at the moment this is experimental API. It depends on some experimental settings that might change in the future.
+	 * The claim about clientId being changed simulteniously with raising "connected" event and being up-to-date only valid when
+	 * this API is consumed at loader / hosing layer. Container Runtime layer will get these changes delayed, as loader layer changes
+	 * propagate slowly through the system, and thus newer runtimes will continue to observe old (non-synchronized) behavior when paired
+	 * with old loader
+	 *
+	 * While it's marked as experimental, this promise could be broken, and consumers  could experience current clientId being changed
+	 * (and "clientIdChanged" event fired) while (only applicable for "read" kind of connections)
+	 * 1. Such clientId is not present in Audience
+	 * 2. Client is not fully caught up
+	 *
+	 * @experimental
+	 */
+	readonly currentClientId: string | undefined;
```

Comment:
If we don't want this to become `undefined` when a connected client disconnects from the ordering service, it seems preferable to call it `latestClientId` and document very explicitly that it'll be undefined if the client has never connected, but then it'll just be the most recent one even if the client becomes disconnected.

Reply:
I do not think "latest" is the right word here to use. At least none of our existing API surfaces uses this word. We use plain "clientId" all over the place. I'd use it here, but I do not like the fact that getMember's argument is also called clientId - that makes it a bit confusing on relationship. It's "current" in a sense that it is only loosely coupled with connections / disconnections - even when we reconnected, we keep old clientId for a while - as long as there is a chance to receive an op with such clientId ("local" op). Mayabe one can call it "latest", but I think "current" is more relevant here.

Reply:
I'm not sure what you mean by "internally". I think you mean - each connection has clientId. But connections are not exposed in our programming model. Only thing we expose is Container class managed clientId. And they change according to rules of Container class, somewhat disconnected from connections themselves.

Reply:
Right, "connections are not exposed in our programming model" is basically what I meant. I wanted to confirm that once a new connection is established, this property will immediately have the clientId of that one and users of the container can't get to the old one anymore (regardless of what the container does or doesn't do with that old clientId "internally", i.e. without letting the user know about it).

Reply:
> this property will immediately have the clientId of that one 
It's not immediate. clientId changes (as observed through public APIs) only after we caught up with ops. It's even more complicated (in some cases we wait for "leave" op from old connection). But it all does not matter - there is no 1:1 relationship here. Connections are "virtualized", and for most part users should not know about them. There are exceptions (signals is the most strong one - we allow sending signals only when there is a connection).

Reply:
Not sure what to do here :)
On one hand, I hear you RE "latest" word. On another, nowhere in our code / comments / documentation we talk about it "latest", we talk about it as "current". Maybe if we use word "self" here (either directly on this property, or wrap with an extra object that has clientId property), it gets clearer (or rather - does not use words that create confusion)?

Reply:
Yeah, "self' seems to skirt around the potential confusion of what it means when the client is connected VS not. In the broader context of "how can we make Audience less confusing/ambiguous with regards to connectivity state", that seems like a good choice. That said, specifically for clientId I think keeping "current" might be fine, if it's well documented that if the client becomes disconnected, it retains its "selfness" (from its perspective it's the same client, after all) and thus its clientId.

My bigger pet peeve is with the contract for getting members (other than the client itself) out of an Audience while the client is disconnected, and what that does/should-do; but that's not for this PR to address :), just context.

Reply:
let me describe what I see in the code RE when members are added/removed from audience.
First, it only applies to "read" connections, and we are trying really hard not to talk too much about differences between "read" and "write" connections (and most of our APIs do not expose that detail).
When we establish new connection, we clear all old clients and immediately populate with new clients - it happens synchronously. But "we establish new connection" is not visible otherwise to a user in any form, as we will raise "connected" event (and change self clientId) only when we are caught up.

And "write" clients get updated as we process ops, including when we have no socket connection (by fetching ops from storage).

I do not see an easy way to tell all that in easily consumable form to the user. I think the best we can say - once "connected" event fires, you have new Audience. And that users can observe all changes through events.

Reply:
You're certainly managing to convey out all the complexities and I agree "it's complicated" might be very close to what we can say about the audience in relation to connection state 😄 .

> And "write" clients get updated as we process ops, including when we have no socket connection (by fetching ops from storage).

Ah, that's a really good case that wasn't obvious to me. "Disconnected" doesn't necessarily mean "completely offline" it just means "no delta stream", correct? In that scenario, are we periodically fetching ops from storage? Or is it something that happens at certain points but not again afterwards? If it happens periodically, then I might indeed consider the Audience still "functional". Both joining and leaving clients would eventually be reflected there for the disconnected client, maybe just with extra delay (but there could be delay even when connected, so eh).

If a disconnected client _doesn't_ fetch ops from storage periodically, then the audience does become stale, doesn't it? A client that left would never disappear from the audience of the disconnected client, and a client that joined would never show up? That's the kind of scenario I was trying to clarify for code that uses Audience. It seems to depend on whether something is fetching ops from storage periodically while a client is disconnected.

Reply:
It's async system, network calls could be issued long time ago and complete at any point in time. If client needs ops, it will bang on storage until it gets them, it might get tail of ops, and it could happen at any time. We also ask for ops proactively while reconnecting. So ideally, we should not even talk in "connected" terms, but rather in terms of "it's good time to send ops", "it's good time to send signals", etc. But there will always be scenarios where we have no good alternative to "you just need to know if socket is connected".  And to some extent, "clientId" is the same bucket - we have to have it as we have no better tools in this space.

Yes, audience is stale on disconnect. It has to do with both ops and signals ("write" and "read" clients).

That said, take a look at the latest commit. I've put more English into it. It does not cover all these topics (as I think it's a bit too much), but it does provide some insights into how to use the system / what users can rely on.